src/combinatorics/simple_graph/triangle/basic.lean

```diff
@@ -28,21 +33,183 @@ This module defines and proves properties about triangles in simple graphs.
 
-open finset fintype nat
-open_locale classical
+namespace finset
+variables {α : Type*} [decidable_eq α]
+
+lemma insert_comm (a b : α) (s : finset α) : insert a (insert b s) = insert b (insert a s) :=
+coe_injective $ by { push_cast, exact set.insert_comm _ _ _ }
+
+end finset
+
+namespace simple_graph
+variables {V : Type*}
+
+@[simp] lemma edge_set_top : (⊤ : simple_graph V).edge_set = {e : sym2 V | ¬ e.is_diag} :=
+by ext x; induction x using sym2.ind; simp
+
+@[simp] lemma edge_finset_top [fintype V] [decidable_eq V] [fintype (⊤ : simple_graph V).edge_set] :
+  (⊤ : simple_graph V).edge_finset = finset.univ.filter (λ e, ¬ e.is_diag) :=
+by ext x; induction x using sym2.ind; simp
+
+end simple_graph
+
+open finset fintype (card) nat
 
 namespace simple_graph
-variables {α 𝕜 : Type*} [fintype α] [linear_ordered_field 𝕜] {G H : simple_graph α} {ε δ : 𝕜}
-  {n : ℕ} {s : finset α}
+variables {α β 𝕜 : Type*} [linear_ordered_field 𝕜] {G H : simple_graph α} {ε δ : 𝕜} {n : ℕ}
+  {s : finset α}
+
+section locally_linear
+
+/-- A graph has edge-disjoint triangles if each edge belongs to at most one triangle. -/
+def edge_disjoint_triangles (G : simple_graph α) : Prop :=
+(G.clique_set 3).pairwise $ λ x y, (x ∩ y : set α).subsingleton
+
+/-- A graph is locally linear if each edge belongs to exactly one triangle. -/
+def locally_linear (G : simple_graph α) : Prop :=
+G.edge_disjoint_triangles ∧ ∀ ⦃x y⦄, G.adj x y → ∃ s, G.is_n_clique 3 s ∧ x ∈ s ∧ y ∈ s
+
+protected lemma locally_linear.edge_disjoint_triangles :
+  G.locally_linear → G.edge_disjoint_triangles :=
+and.left
+
+lemma edge_disjoint_triangles.mono (h : G ≤ H) (hH : H.edge_disjoint_triangles) :
+  G.edge_disjoint_triangles :=
+hH.mono $ clique_set_mono h
+
+@[simp] lemma edge_disjoint_triangles_bot : (⊥ : simple_graph α).edge_disjoint_triangles :=
+by simp [edge_disjoint_triangles]
+
+@[simp] lemma locally_linear_bot : (⊥ : simple_graph α).locally_linear := by simp [locally_linear]
+
+lemma edge_disjoint_triangles.map (f : α ↪ β) (hG : G.edge_disjoint_triangles) :
+  (G.map f).edge_disjoint_triangles :=
+begin
+  rw [edge_disjoint_triangles, clique_set_map (bit1_lt_bit1.2 zero_lt_one),
+    ((finset.map_injective f).inj_on _).pairwise_image],
+  classical,
+  rintro s hs t ht hst,
+  dsimp [function.on_fun],
+  rw [←coe_inter, ←map_inter, coe_map, coe_inter],
+  exact (hG hs ht hst).image _,
+end
+
+lemma locally_linear.map (f : α ↪ β) (hG : G.locally_linear) : (G.map f).locally_linear :=
+begin
+  refine ⟨hG.1.map _, _⟩,
+  rintro _ _ ⟨a, b, h, rfl, rfl⟩,
+  obtain ⟨s, hs, ha, hb⟩ := hG.2 h,
+  exact ⟨s.map f, hs.map, mem_map_of_mem _ ha, mem_map_of_mem _ hb⟩,
+end
+
+@[simp] lemma locally_linear_comap {G : simple_graph β} {e : α ≃ β} :
+  (G.comap e).locally_linear ↔ G.locally_linear :=
+begin
+  refine ⟨λ h, _, _⟩,
+  { rw [←comap_map_eq e.symm.to_embedding G, comap_symm, map_symm],
+    exact h.map _ },
+  { rw [←equiv.coe_to_embedding, ←map_symm],
+    exact locally_linear.map _ }
+end
```

Comment:
Is one direction of this true for embeddings?

Reply:
This comment belongs to #19201.

I cannot really work on it because I need #19203 to be in, but I suspect the only implication that can be weakened is `G.locally_linear → (G.map f).locally_linear`.